python/src/iceberg/io/base.py

```diff
@@ -0,0 +1,123 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from abc import ABC, abstractmethod
+
+
+class InputFile(ABC):
+    """A base class for InputFile implementations"""
+
+    def __init__(self, location: str):
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the input file"""
+        return self._location
+
+    @property
+    @abstractmethod
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
+
+    @abstractmethod
+    def __enter__(self):
+        """Enter context for InputFile
+
+        This method should assign a seekable stream to `self.input_stream` and
+        return `self`. If the file does not exist, a FileNotFoundError should
+        be raised."""
+
+    @abstractmethod
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Exit context for InputFile
+
+        This method should perform any necessary teardown."""
+
+
+class OutputFile(ABC):
+    """A base class for OutputFile implementations"""
+
+    def __init__(self, location: str, overwrite: bool = False):
```

Comment:
Since this is a stream factory, overwrite is actually an option for the output stream and not for the `OutputFile` itself. Using a method to open like I suggested for `InputFile` gives you a place to put this:

```python
with out_file.create(overwrite=True) as f:
    f.write(...)

in_file = out_file.to_input_file()
```

Reply:
Fixed! Thanks @rdblue